holidayextras/make-up
# Allow filtering of files to lint by changes between git branches

#### What does this PR do? (please provide any background)
Allow filtering of files to lint by changes between git branches. We want to lint our files automatically when we push to GitHub, but we're getting annoyed by how long it takes to lint the whole tree of a big project and just bypass the hook most of the time.

#### What tests does this PR have?
New unit tests in:
* `test/lib/gitBranchFilter.js`
```
gitBranchFilter
  process
    when `git.currentBranch` errors
      ✓ callsback with error
    when `git.nameOnlyDiff` errors
      ✓ callsback with error
    when `git` functions succeed
      ✓ calls `git.nameOnlyDiff` with the current branch
      ✓ adds root level slashes to `git.nameOnlyDiff` result
      ✓ should callback with intersection of files and `git.nameOnlyDiff` result
```
* `test/lib/integrations/eslint.js`
```
_getFilters()
  with null options
    ✓ returns an empty filters list
  with empty options
    ✓ returns an empty filters list
  with a single option
    ✓ returns an array with a single filter function
  with two options
    ✓ returns an array with two filter functions
_filterFilenames()
  with null filters
    ✓ returns unmodified files
  with empty filters
    ✓ returns unmodified files
  with filters
    ✓ applies chain of filters correctly
_processGlobs
  with some files
    with _checkFiles
      ✓ calls since filter if enabled by option
      ✓ calls git since filter if enabled by option
      ✓ calls git branch filter if enabled by option
      ✓ calls all enabled filters
```
* test/lib/util.js
```
addRootLevelSlashes()
  with a root level file
    ✓ adds slashes
  with a nested file
    ✓ does not change it
```
#### How can this be tested?
Checkout this branch and run:
```
./bin/make-up.js -b origin/master bin lib test
```
verify that only the files modified in the PR are linted and compare the differences with plain `make-up`:
```
./bin/make-up.js bin lib test
```

#### Screenshots / Screencast
N/A

#### What gif best describes how you feel about this work?
![Monty Python Flying Mustache Head](http://i.giphy.com/XJKW9YWXyz0MU.gif)

---


#### Reviewers

**Review 1** (HX Distribution)

**Review 2** (Short Breaks)

**Review 3** _(optional)_

By adding a +1 you are confirming you have...
- Witnessed the work behaving as expected (this could be on the author's machine or screencast).
- Checked for coding anti-patterns.
- Checked for appropriate test coverage.
- Checked all the tests are passing.

\*  for HX this review must be completed by an SE, SA or Project Guru
